service/history/historyEngine.go

```diff
@@ -2335,16 +2411,24 @@ func (e *historyEngineImpl) updateWorkflowExecution(
 ) error {
 
 	return e.updateWorkflowExecutionWithAction(ctx, domainID, execution,
-		func(mutableState mutableState) (*updateWorkflowAction, error) {
-			err := action(mutableState)
-			if err != nil {
-				return nil, err
-			}
-			postActions := &updateWorkflowAction{
-				createDecision: createDecisionTask,
-			}
-			return postActions, nil
-		})
+		getUpdateWorkflowActionFunc(createDecisionTask, action))
+}
+
+func getUpdateWorkflowActionFunc(
+	createDecisionTask bool,
+	action func(builder mutableState) error,
+) updateWorkflowActionFunc {
+
+	return func(builder mutableState) (*updateWorkflowAction, error) {
+		err := action(builder)
+		if err != nil {
+			return nil, err
+		}
+		postActions := &updateWorkflowAction{
+			createDecision: createDecisionTask,
+		}
+		return postActions, nil
+	}
```

Comment:
let us get rid of this function and create 2 vars and let caller just return these 2 vars
```
var updateWorkflowWithNewDecision = updateWorkflowAction{
	createDecision: true,
}
var updateWorkflowWithoutDecision = updateWorkflowAction{
	createDecision: false,
}
```

Reply:
one more thing, plz change *updateWorkflowAction to updateWorkflowAction, i.e. pointer to struct. since there is no case that updateWorkflowAction can be nil

Reply:
Updated